### DISCUSSION TITLE
Odt-based exporter for a French magazine named GNU/Linux Magazine.

### DISCUSSION THREAD

**Discussion by kjambunathan** [kjambunathan/org-mode-ox-odt#213] [General]
@DamienCassou 

> From [Need for dedicated kinds of paragraphs](https://lists.gnu.org/archive/html/emacs-orgmode/2022-10/msg00559.html)

> I'm implementing an odt-based exporter for a French magazine named
GNU/Linux Magazine.  This magazine defines several kinds of boxes, which
are small paragraphs of a certain type among "Default", "Attention",
"Warning" and "PAO". When published, the magazine will change the
background of these boxes depending on the type (e.g., using red color
for warning boxes).

Try my fork of ODT exporter.  The solution you want for your FOSS articles already exists in my fork.

Here is how you would solve the problem.

To get this

![Screenshot from 2022-10-18 07-43-13](https://user-images.githubusercontent.com/19869864/196319463-d3e0fb67-a9f8-4bb8-85ed-eda1d3480686.png)

do this

```
#+ATTR_ODT: :target "extra_styles"
#+begin_src nxml
<style:style style:name="CustomParagraph"
             style:parent-style-name="Standard"
             style:family="paragraph"
             style:class="text">
  <style:paragraph-properties fo:margin-bottom="0.212cm"
                              fo:margin-top="0cm" />
  <style:text-properties fo:color="#ff0000" />
</style:style>
#+end_src

#+ATTR_ODT: :target "extra_styles"
#+begin_src nxml
<style:style style:name="CustomTextBox"
             style:parent-style-name="Frame"
             style:family="graphic">
  <style:graphic-properties draw:fill="solid"
                            draw:fill-color="#ffff00"
                            draw:wrap-influence-on-position="once-concurrent"
                            fo:background-color="#ffff00"
                            style:horizontal-pos="center"
                            style:horizontal-rel="paragraph-content"
                            style:vertical-pos="top"
                            style:vertical-rel="paragraph"
                            style:wrap="none"
                            svg:x="0cm"
                            svg:y="0cm"
                            text:anchor-type="char" />
</style:style>
#+end_src

#+ATTR_ODT: :style  "CustomParagraph"
Officia sunt cupidatat cupidatat enim do cupidatat do ea excepteur
eiusmod non mollit labore nostrud sit laboris aliqua consequat.  Amet,
commodo labore nisi commodo eiusmod irure magna.  Consectetur ea in
aliqua velit laboris dolore consequat.  Culpa dolor aliqua in
proident, pariatur ea amet, reprehenderit nostrud cupidatat non
consequat culpa ipsum.  Exercitation et proident, cillum elit, ut ea
et ullamco quis incididunt fugiat incididunt ad nostrud excepteur.

#+ATTR_ODT: :style "CustomTextBox"
#+BEGIN_textbox
Aliqua reprehenderit mollit ad sit voluptate id pariatur.  Consequat
occaecat minim officia eiusmod aliquip qui lorem dolore deserunt
laboris aliquip eu sed fugiat.
#+END_textbox
```
See the README of this project.    The issues page has plenty of screenshots and documentation.  They will inform you of the capabilities of the exporter.

I am the original author of the ODT exporter, and I no longer contribute to  or participate in Emacs forums or codebases because of ideological differences.  Apart from ideological differences with FSF,  let me remind you  that the  code in this fork is still GPL--I hold the copyright, and the copyright to new code is not assigned to FSF.

If you are NOT doing a pro-bono work, I would appreciate if you could mention my work to the magazine team, and point them in my direction. I would appreciate some extra income.

**Comment by kjambunathan** [comment]
If you are following https://planet.emacslife.com/, someone posted about [Use `org-extra-emphasis` , when you need more Emphasis Markers in Emacs Org mode](https://emacsnotes.wordpress.com/2022/06/29/use-org-extra-emphasis-when-you-need-more-emphasis-markers-in-emacs-org-mode/).  Btw, I am the author of [`org-extra-emphasis`](https://github.com/QiangF/org-extra-emphasis), which is a dirty hack (but gets the job done).  No, I am NOT the author of that blog.

**Comment by kjambunathan** [comment]
I wouldn't hesitate to say this ...The people who are suggesting solutions to you on your question, have no clue about the ODT backend, or the peculiarities of OpenDocument format.    Their suggestions are NOT informed by their experience.

Even if you dislike my "in the face" comments here, you will be off to a better start by listening to my suggestion here, and roll out your own  code.

**Comment by DamienCassou** [comment]
Thank you for your help.

**Comment by kjambunathan** [comment]
If they have an existing template `.ott` file, and if you don't mind sharing it, I would be interested in taking a look.

If you are the one creating the template file, export your file with 

```
#+odt_preferred_output_format: ott
```
and assuming that `LibreOffice` is installed and `soffice` is in your path, you will have a `.ott` template file.

These are the `odt` specific keywords, that are available to you.  You can insert them with `C-c C-e # o`.  These should be a good starting point.  Start exploring with keywords that have `styles` in their names.

```
#+options: tex:t
#+description:
#+keywords:
#+subtitle:
#+odt_math_syntax: latex
#+odt_preferred_output_format: ott
#+odt_styles_file:
#+odt_extra_images:
#+odt_extra_styles:
#+odt_extra_automatic_styles:
#+odt_master_styles:
#+odt_content_template_file:
#+odt_automatic_styles:
#+odt_display_outline_level: 2
#+odt_app:
#+odt_document_properties:
#+odt_extra_meta:
#+odt_prettify_xml:
#+odt_validate:
#+odt_endnote_regexp: ^$
````

**Reply by DamienCassou** [reply]
> If they have an existing template .ott file, and if you don't mind sharing it, I would be interested in taking a look.

sure: https://github.com/GLMF/outils_auteurs/raw/master/gnulinuxmagazine.ott

**Reply by kjambunathan** [reply]
> > If they have an existing template .ott file, and if you don't mind sharing it, I would be interested in taking a look.
> 
> sure: https://github.com/GLMF/outils_auteurs/raw/master/gnulinuxmagazine.ott

If you are contributing content to their article, and if you would like to compose your article in `org` mode,  the attached `zip` file tells you how the `org` content will look like.

The sample article uses the following `char style`-s

```lisp
("url" "menu" "italic" "gras" "code_5f_par" "code_5f_em")
```
and they don't map directly to one of the standard `org` emphasis markers.   So, I have made all the `span` elements to `bold`.   With `elisp` glue code, one can map  `org bold` style to map to say `gras` , `org verbatim` style to map to say `code_5f_em` etc. 

The `zip` file has a `odt` file which is created by this exporter.  You can satisfy yourself that most of the magazine-defined styles carries over.  Note the use of 

```
#+odt_styles_file: ./styles.xml
```
which tells the ODT exporter that style definition comes from the GNU Linux Magazine.


[discussion213.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/9820019/discussion213.zip)


```
M    Mode     Size        Date&time         Filename
- ----------  -----  --------------------  ----------------
  drwxr-xr-x      0  19-Oct-2022 17:01:42  discussion213/
  -rw-r--r--  24565  19-Oct-2022 17:00:20  discussion213/styles.xml
  -rw-r--r--  21136  19-Oct-2022 17:00:58  discussion213/gnulinuxmagazine-draft.odt
  -rw-r--r--  13953  19-Oct-2022 17:00:52  discussion213/gnulinuxmagazine-draft.org
- ----------  -----  --------------------  ----------------
              59654                         4 files
```

**Comment by kjambunathan** [comment]
This is how the contents in the `zip` file looks like on my end.  Everything is machine-generated, so I haven't looked at the contents of the template file.  For a first cut, I believe it takes you very far. 

**_`org` to `odt` but with GNU Linux Magazine defined styles_**

![Screenshot from 2022-10-19 17-46-58](https://user-images.githubusercontent.com/19869864/196692799-ff1ba6b8-084a-4d04-9fcd-cf3cf572e431.png)

**Comment by kjambunathan** [comment]
Instead of replying to you, I think I ended up editing your message.  You may not have received notifications for my edits.  Take a look at https://github.com/kjambunathan/org-mode-ox-odt/discussions/213#discussioncomment-3914557 and also the screenshot.

**Reply by DamienCassou** [reply]
Thank you very much for your brilliant job!

**Comment by kjambunathan** [comment]
> Thank you very much for your brilliant job!

I wasn't looking for any motivational speech from your side.  No need  to be snarky.

I have locked this conversation.
